### ISSUE TITLE
[stable/jenkins] Cannot import Jobs

### ISSUE THREAD

**Issue by azbpa** [helm/charts#23453]
**Describe the bug**
I am trying to import jobs using the `master.jobs` value. However, it is not working. After deployment, the jobs folder is totally empty in the filesystem of the Jenkins Pod.

Looking at the Charts implementation, it seems like the jobs should be stored as config maps that get mounted to the init-container for further processing. In my deployment I can see the config map being created, however, looking at my `Deployment` spec it is not defined as volume and it is also not mounted to the path `/var/jenkins_jobs` inside the container.

The behavior seems strange to me since the logic (`{{- if .Values.master.jobs }}`) for creating the config map and mounting them  is similar:
https://github.com/helm/charts/search?q=%7B%7B-+if+.Values.master.jobs+%7D%7D&unscoped_q=%7B%7B-+if+.Values.master.jobs+%7D%7D

**Version of Helm and Kubernetes**:
Helm: 3.2.4
Kubernetes 1.18.6

**Which chart**:
stable/jenkins 2.5.0

**How to reproduce it** (as minimally and precisely as possible):

A minimal example would be something like this (it is no valid job xml file, but it already shows that the volume and mount are missing in the `Deployment` yaml):
```
helm template myrelease stable/jenkins --set master.jobs.test-job="<xml>...</xml>"
```
Output:
```yaml
---
# Source: jenkins/templates/service-account.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: myrelease-jenkins
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
---
# Source: jenkins/templates/secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: myrelease-jenkins
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
type: Opaque
data:
  jenkins-admin-password: "b1ZFMExNRW13Uw=="
  jenkins-admin-user: "YWRtaW4="
---
# Source: jenkins/templates/config.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: myrelease-jenkins
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
data:
  apply_config.sh: |-
    echo "applying Jenkins configuration"
    mkdir -p /usr/share/jenkins/ref/secrets/;
    echo "disable Setup Wizard"
    # Prevent Setup Wizard when JCasC is enabled
    echo $JENKINS_VERSION > /var/jenkins_home/jenkins.install.UpgradeWizard.state
    echo $JENKINS_VERSION > /var/jenkins_home/jenkins.install.InstallUtil.lastExecVersion
    echo "download plugins"
    # Install missing plugins
    cp /var/jenkins_config/plugins.txt /var/jenkins_home;
    rm -rf /usr/share/jenkins/ref/plugins/*.lock
    /usr/local/bin/install-plugins.sh `echo $(cat /var/jenkins_home/plugins.txt)`;
    echo "copy plugins to shared volume"
    # Copy plugins to shared volume
    yes n | cp -i /usr/share/jenkins/ref/plugins/* /var/jenkins_plugins/;
    echo "finished initialization"
  plugins.txt: |-
    kubernetes:1.25.7
    workflow-job:2.39
    workflow-aggregator:2.6
    credentials-binding:1.23
    git:4.2.2
    configuration-as-code:1.41
---
# Source: jenkins/templates/jcasc-config.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: myrelease-jenkins-jenkins-jcasc-config
  namespace: default
  labels:
    "app.kubernetes.io/name": jenkins
    "helm.sh/chart": jenkins-2.5.0
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
    myrelease-jenkins-jenkins-config: "true"
data:
  jcasc-default-config.yaml: |-
    jenkins:
      authorizationStrategy:
        loggedInUsersCanDoAnything:
          allowAnonymousRead: false
      securityRealm:
        legacy
      disableRememberMe: false
      remotingSecurity:
        enabled: true
      mode: NORMAL
      numExecutors: 0
      projectNamingStrategy: "standard"
      markupFormatter:
        plainText
      clouds:
      - kubernetes:
          containerCapStr: "10"
          defaultsProviderTemplate: ""
          connectTimeout: "5"
          readTimeout: "15"
          jenkinsUrl: "http://myrelease-jenkins:8080"
          jenkinsTunnel: "myrelease-jenkins-agent:50000"
          maxRequestsPerHostStr: "32"
          name: "kubernetes"
          namespace: "default"
          serverUrl: "https://kubernetes.default"
          podLabels:
          - key: "jenkins/myrelease-jenkins-slave"
            value: "true"
          templates:
            - name: "default"
              containers:
              - name: "jnlp"
                alwaysPullImage: false
                args: "^${computer.jnlpmac} ^${computer.name}"
                envVars:
                - containerEnvVar:
                    key: "JENKINS_URL"
                    value: "http://myrelease-jenkins.default.svc.cluster.local:8080"
                image: "jenkins/inbound-agent:4.3-4"
                privileged: "false"
                resourceLimitCpu: 512m
                resourceLimitMemory: 512Mi
                resourceRequestCpu: 512m
                resourceRequestMemory: 512Mi
                runAsUser: 
                runAsGroup: 
                ttyEnabled: false
                workingDir: /home/jenkins
              idleMinutes: 0
              instanceCap: 2147483647
              label: "myrelease-jenkins-slave "
              nodeUsageMode: "NORMAL"
              podRetention: Never
              showRawYaml: true
              serviceAccount: "default"
              slaveConnectTimeoutStr: "100"
              yamlMergeStrategy: override
      crumbIssuer:
        standard:
          excludeClientIPFromCrumb: true
    security:
      apiToken:
        creationOfLegacyTokenEnabled: false
        tokenGenerationOnCreationEnabled: false
        usageStatisticsEnabled: true
    unclassified:
      location:
        adminAddress: 
        url: http://myrelease-jenkins:8080
---
# Source: jenkins/templates/jobs.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: myrelease-jenkins-jobs
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
data:
  test-job: <xml>...</xml>
---
# Source: jenkins/templates/tests/test-config.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: myrelease-jenkins-tests
  namespace: default
data:
  run.sh: |-
    @test "Testing Jenkins UI is accessible" {
      curl --retry 48 --retry-delay 10 myrelease-jenkins:8080/login
    }
---
# Source: jenkins/templates/home-pvc.yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myrelease-jenkins
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
spec:
  accessModes:
    - "ReadWriteOnce"
  resources:
    requests:
      storage: "8Gi"
---
# Source: jenkins/templates/rbac.yaml
# This role is used to allow Jenkins scheduling of agents via Kubernetes plugin.
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: myrelease-jenkins-schedule-agents
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
rules:
- apiGroups: [""]
  resources: ["pods", "pods/exec", "pods/log", "persistentvolumeclaims", "events"]
  verbs: ["get", "list", "watch"]
- apiGroups: [""]
  resources: ["pods", "pods/exec", "persistentvolumeclaims"]
  verbs: ["create", "delete", "deletecollection", "patch", "update"]
---
# Source: jenkins/templates/rbac.yaml
# The sidecar container which is responsible for reloading configuration changes
# needs permissions to watch ConfigMaps
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: myrelease-jenkins-casc-reload
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
rules:
- apiGroups: [""]
  resources: ["configmaps"]
  verbs: ["get", "watch", "list"]
---
# Source: jenkins/templates/rbac.yaml
# We bind the role to the Jenkins service account. The role binding is created in the namespace
# where the agents are supposed to run.
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: myrelease-jenkins-schedule-agents
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: myrelease-jenkins-schedule-agents
subjects:
- kind: ServiceAccount
  name: myrelease-jenkins
  namespace: default
---
# Source: jenkins/templates/rbac.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: myrelease-jenkins-watch-configmaps
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: myrelease-jenkins-casc-reload
subjects:
- kind: ServiceAccount
  name: myrelease-jenkins
  namespace: default
---
# Source: jenkins/templates/jenkins-agent-svc.yaml
apiVersion: v1
kind: Service
metadata:
  name: myrelease-jenkins-agent
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
spec:
  ports:
    - port: 50000
      targetPort: 50000
      
      name: slavelistener
  selector:
    "app.kubernetes.io/component": "jenkins-master"
    "app.kubernetes.io/instance": "myrelease"
  type: ClusterIP
---
# Source: jenkins/templates/jenkins-master-svc.yaml
apiVersion: v1
kind: Service
metadata:
  name: myrelease-jenkins
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
spec:
  ports:
    - port: 8080
      name: http
      targetPort: 8080
  selector:
    "app.kubernetes.io/component": "jenkins-master"
    "app.kubernetes.io/instance": "myrelease"
  type: ClusterIP
---
# Source: jenkins/templates/jenkins-master-deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myrelease-jenkins
  namespace: default
  labels:
    "app.kubernetes.io/name": 'jenkins'
    "helm.sh/chart": "jenkins-2.5.0"
    "app.kubernetes.io/managed-by": "Helm"
    "app.kubernetes.io/instance": "myrelease"
    "app.kubernetes.io/component": "jenkins-master"
spec:
  replicas: 1
  strategy:
    type: Recreate
  selector:
    matchLabels:
      "app.kubernetes.io/component": "jenkins-master"
      "app.kubernetes.io/instance": "myrelease"
  template:
    metadata:
      labels:
        "app.kubernetes.io/name": 'jenkins'
        "helm.sh/chart": "jenkins-2.5.0"
        "app.kubernetes.io/managed-by": "Helm"
        "app.kubernetes.io/instance": "myrelease"
        "app.kubernetes.io/component": "jenkins-master"
      annotations:
        checksum/config: 51aa3ca7a7643f4c64a855eceb613fe54a3a161cbdc2a91cb739de378c03dd74
    spec:
      securityContext:
        runAsUser: 1000
        fsGroup: 1000
      serviceAccountName: "myrelease-jenkins"
      initContainers:
        - name: "copy-default-config"
          image: "jenkins/jenkins:lts"
          imagePullPolicy: "Always"
          command: [ "sh", "/var/jenkins_config/apply_config.sh" ]
          env:
            - name: ADMIN_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: myrelease-jenkins
                  key: jenkins-admin-password
            - name: ADMIN_USER
              valueFrom:
                secretKeyRef:
                  name: myrelease-jenkins
                  key: jenkins-admin-user
          resources:
            limits:
              cpu: 2000m
              memory: 4096Mi
            requests:
              cpu: 50m
              memory: 256Mi
          volumeMounts:
            - mountPath: /tmp
              name: tmp
            - mountPath: /var/jenkins_home
              name: jenkins-home
            - mountPath: /var/jenkins_config
              name: jenkins-config
            - mountPath: /usr/share/jenkins/ref/secrets/
              name: secrets-dir
            - mountPath: /usr/share/jenkins/ref/plugins
              name: plugins
            - mountPath: /var/jenkins_plugins
              name: plugin-dir
      containers:
        - name: jenkins
          image: "jenkins/jenkins:lts"
          imagePullPolicy: "Always"
          args: [ "--argumentsRealm.passwd.$(ADMIN_USER)=$(ADMIN_PASSWORD)",  "--argumentsRealm.roles.$(ADMIN_USER)=admin", "--httpPort=8080"]
          env:
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: JAVA_OPTS
              value: >
                
                 -Dcasc.reload.token=$(POD_NAME) 
            - name: JENKINS_OPTS
              value: ""
            - name: JENKINS_SLAVE_AGENT_PORT
              value: "50000"
            - name: ADMIN_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: myrelease-jenkins
                  key: jenkins-admin-password
            - name: ADMIN_USER
              valueFrom:
                secretKeyRef:
                  name: myrelease-jenkins
                  key: jenkins-admin-user
            - name: CASC_JENKINS_CONFIG
              value: /var/jenkins_home/casc_configs
          ports:
            - containerPort: 8080
              name: http
            - containerPort: 50000
              name: slavelistener
          livenessProbe:
            httpGet:
              path: "/login"
              port: http
            initialDelaySeconds: 90
            periodSeconds: 10
            timeoutSeconds: 5
            failureThreshold: 5
          readinessProbe:
            httpGet:
              path: "/login"
              port: http
            initialDelaySeconds: 60
            periodSeconds: 10
            timeoutSeconds: 5
            failureThreshold: 3

          resources:
            limits:
              cpu: 2000m
              memory: 4096Mi
            requests:
              cpu: 50m
              memory: 256Mi
          volumeMounts:
            - mountPath: /tmp
              name: tmp
            - mountPath: /var/jenkins_home
              name: jenkins-home
              readOnly: false
            - mountPath: /var/jenkins_config
              name: jenkins-config
              readOnly: true
            - mountPath: /usr/share/jenkins/ref/secrets/
              name: secrets-dir
              readOnly: false
            - mountPath: /usr/share/jenkins/ref/plugins/
              name: plugin-dir
              readOnly: false
            - name: sc-config-volume
              mountPath: /var/jenkins_home/casc_configs
        - name: jenkins-sc-config
          image: "kiwigrid/k8s-sidecar:0.1.144"
          imagePullPolicy: IfNotPresent
          env:
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: LABEL
              value: "myrelease-jenkins-jenkins-config"
            - name: FOLDER
              value: "/var/jenkins_home/casc_configs"
            - name: NAMESPACE
              value: "default"
            - name: REQ_URL
              value: "http://localhost:8080/reload-configuration-as-code/?casc-reload-token=$(POD_NAME)"
            - name: REQ_METHOD
              value: "POST"
            - name: REQ_RETRY_CONNECT
              value: "10"
          resources:
            {}
          volumeMounts:
            - name: sc-config-volume
              mountPath: "/var/jenkins_home/casc_configs"
            - name: jenkins-home
              mountPath: /var/jenkins_home

      volumes:
      - name: plugins
        emptyDir: {}
      - name: tmp
        emptyDir: {}
      - name: jenkins-config
        configMap:
          name: myrelease-jenkins
      - name: secrets-dir
        emptyDir: {}
      - name: plugin-dir
        emptyDir: {}
      - name: jenkins-home
        persistentVolumeClaim:
          claimName: myrelease-jenkins
      - name: sc-config-volume
        emptyDir: {}
---
# Source: jenkins/templates/tests/jenkins-test.yaml
apiVersion: v1
kind: Pod
metadata:
  name: "myrelease-ui-test-iej1v"
  namespace: default
  annotations:
    "helm.sh/hook": test-success
spec:
  initContainers:
    - name: "test-framework"
      image: "dduportal/bats:0.4.0"
      command:
      - "bash"
      - "-c"
      - |
        set -ex
        # copy bats to tools dir
        cp -R /usr/local/libexec/ /tools/bats/
      volumeMounts:
      - mountPath: /tools
        name: tools
  containers:
    - name: myrelease-ui-test
      image: jenkins/jenkins:lts
      command: ["/tools/bats/bats", "-t", "/tests/run.sh"]
      volumeMounts:
      - mountPath: /tests
        name: tests
        readOnly: true
      - mountPath: /tools
        name: tools
  volumes:
  - name: tests
    configMap:
      name: myrelease-jenkins-tests
  - name: tools
    emptyDir: {}
  restartPolicy: Never
```
**Anything else we need to know**:

**Comment by DPD-30** [comment]
Having the exact same issue.  went back to basic values.yaml with example job from readme.md same results. Not jobs created.

**Comment by canozyurt** [comment]
According to templates/jenkins-master-deployment.yaml:

```
            {{- if .Values.master.enableXmlConfig }}
            {{- if .Values.master.credentialsXmlSecret }}
            - mountPath: /var/jenkins_credentials
              name: jenkins-credentials
              readOnly: true
            {{- end }}
            {{- if .Values.master.jobs }}
            - mountPath: /var/jenkins_jobs
              name: jenkins-jobs
              readOnly: true
            {{- end }}
            {{- end }}
```
.Values.master.enableXmlConfig should be set to true.

**Comment by azbpa** [comment]
@canozyurt Good hint. Didn't notice that `if` when checking out the file the first time.

Anyways, I might be wrong, but doesn't setting `master.enableXmlConfig` to `true` somehow impact JCasC configuration?

At least it seems to remove the `authorizationStrategy` and `securityRealm` settings from the default JCasC config here: https://github.com/helm/charts/blob/b78b82d3888d3110ac4757d61534d6a5656d4c48/stable/jenkins/templates/_helpers.tpl#L75-L85

It also bumps in a `config.xml` file 
https://github.com/helm/charts/blob/71d46a3af676eeb3f4a41602dc88ecbbe2926bd1/stable/jenkins/templates/config.yaml#L16-L160
and changes the `apply_config.sh` file
https://github.com/helm/charts/blob/71d46a3af676eeb3f4a41602dc88ecbbe2926bd1/stable/jenkins/templates/config.yaml#L170-L196

So, basically I would like to stay with the default JCasC strategy but with the ability to import jobs. I would be fine to use XML, but open in using any other possibility (if there is any). But in any case, I think the import of jobs should be decoupled from the `enableXmlConfig` flag, or do I miss something?

**Comment by canozyurt** [comment]
Appearently it only causes some parameters to be set via xml files instead of JCasC. For example you can find `authorizationStrategy ` and `securityRealm` in config.yaml at lines 29-30. I didn't check the configuration files line by line but my impression is that either way you will have the same outcome.

I also agree with you on decoupling them. I can't see any reason for them to be mutually exclusive.

**Comment by Spritekin** [comment]
+1 

I was having the same problem trying to load jobs and after reading this thread decided to enable the XML config flag and it worked. I agree there should be a specialised flag for loading jobs without requiring CasC.

**Comment by Spritekin** [comment]
I found that disabling the enableXmlConfig flag (and deploy) then enabling it again (and deploy again) will cause the whole service to be reinstalled without losing the current jobs so it works as a refresh.

Is the only way I have been able to reffresh the jobs.

**Comment by scottrigby** [comment]
📢 `stable/jenkins` has moved to https://github.com/jenkinsci/helm-charts. If still applicable, please reopen the issue there. Thanks!
